Workspace: Show text sets for used fonts in story

carloskelly13 opened the pull request:
## Summary

Show text sets that map to currently used fonts.

![image](https://user-images.githubusercontent.com/1738349/96178315-d8f9f100-0ef4-11eb-9b2b-29382b7b127d.png)

No text boxes with `UnifrakturMaguntia` 

![image](https://user-images.githubusercontent.com/1738349/96178344-e44d1c80-0ef4-11eb-8551-f8e697c48875.png)

## Relevant Technical Choices

Uses existing text set filter pattern to show text sets that map to currently used fonts.

## To-do

None

## User-facing changes

None (because feature flag)

Shows a new "Fonts In Use" pill that will show the correct text sets

## Testing Instructions

1. Create a story and add some text boxes.
2. `Open Sans` has a number of text sets and `UnifrakturMaguntia` is pretty unique. 
3. Delete one of the text boxes and see the text sets update.

---

<!-- Please reference the issue(s) this PR addresses. -->

#3640

carloskelly13 committed:
Added tests for finding in-common text sets.

carloskelly13 commented on assets/src/edit-story/components/library/panes/text/textSets/getTextSetReducer.js:
```diff
@@ -19,27 +19,28 @@ export const TEXT_SET_ACTIONS = {
+  { filteredTextSets, renderedTextSets },
+  action
+) {
+  switch (action.type) {
```
This simplifies the reducer function to not be a curried wrapper.

carloskelly13 commented on assets/src/edit-story/components/library/panes/text/textSets/textSets.js:
```diff
@@ -84,25 +90,40 @@ function TextSets() {
+  const storyPages = useStory(({ state: { pages } }) => pages);
+
   const ref = useRef();
+  const inUseFonts = useRef([]);
```
We want to use a ref to store the currently used fonts. If we just make this an on-the-fly computed value, then every single time you move an object or make any change to the editor, it will think things changed. We want to cut down on re-rendering the same filtered text sets.

carloskelly13 commented on assets/src/edit-story/components/library/panes/text/textSets/textSets.js:
```diff
@@ -84,25 +90,40 @@ function TextSets() {
+    const updatedInUseFonts = getInUseFontsForPages(storyPages);
+    if (
+      updatedInUseFonts.length === inUseFonts.current.length &&
+      updatedInUseFonts.every((font, idx) => font === inUseFonts.current[idx])
```
This is the check to ensure that while the array might have changed, let's make sure the values (string font names) are the same.

carloskelly13 commented on assets/src/edit-story/utils/getInUseFonts.js:
```diff
@@ -0,0 +1,57 @@
+ * limitations under the License.
+ */
+
+export function getInUseFontsForPages(pages) {
```
This function enumerates through all the pages and creates a unique array of font names.

swissspidy replied:
On the surface there might be some overlap with `assets/src/edit-story/output/utils/getFontDeclarations.js`. Perhaps there's something that could be shared between the two, or at least referenced in the jsdoc (which are currently missing). Anyhow, just dropping this here

carloskelly13 replied:
Sounds good. I will add some JSDocs. There is some overlap, but we don't need the in-depth nature of getting weights or building the string. The only really in common line is:

`const textElements = elements.filter(({ type }) => type === 'text');`

which is probably not worth extracting out, but I don't feel strongly either way.

carloskelly13 commented on assets/src/edit-story/utils/getInUseFonts.js:
```diff
@@ -0,0 +1,57 @@
+  ];
+}
+
+export function getTextSetsForFonts({ fonts, textSets }) {
```
This function returns all the text sets for a list of font names.

carloskelly13 committed:
Add new function to search pages for in-use fonts with tests.

# Conflicts:
#	assets/src/edit-story/components/library/panes/text/textSets/textSets.js

carloskelly13 committed:
Add In-Use Pill, clean up reducer

# Conflicts:
#	assets/src/edit-story/components/library/panes/text/textSets/getTextSetReducer.js
#	assets/src/edit-story/components/library/panes/text/textSets/textSets.js

carloskelly13 committed:
Render text sets as they are used by the story.

# Conflicts:
#	assets/src/edit-story/components/library/panes/text/textSets/textSets.js

carloskelly13 committed:
Update tests for simplified getTextSetReducer

# Conflicts:
#	assets/src/edit-story/components/library/panes/text/textSets/test/getTextSetReducer.js

carloskelly13 committed:
Update for new virtualization refactor.

carloskelly13 committed:
Added JSDocs for helper functions.

carloskelly13 committed:
Remove old unused files

barklund commented on assets/src/edit-story/utils/getInUseFonts.js:
```diff
@@ -0,0 +1,71 @@
+          ];
+        }, [])
+        .filter(Boolean)
+    ),
```
Would this not be a bit more elegant with `.map()` and `.flat()` rather than `.reduce()` and all those spreads? I also think it'll be more performant, but that's probably on the order of nanos.

carloskelly13 replied:
Oh yeah. Good call, I sometimes forget about that ES19 gem, `.flat()`.

barklund commented on assets/src/edit-story/utils/getInUseFonts.js:
```diff
@@ -0,0 +1,71 @@
+      );
+      return [...textSetMemo, hasFontInUse && currentTextSet];
+    }, [])
+    .filter(Boolean);
```
Here I especially don't see the need for (at least) the outer `.reduce()` - it should just be a `.map()`.

barklund reviewed (approved):
Seems elegant enough and appears to be working.

carloskelly13 committed:
Cleaned up util functions to reduce the usage of reduce.

dmmulroy commented on assets/src/edit-story/components/library/panes/text/textSets/textSets.js:
```diff
@@ -112,16 +132,16 @@ function TextSets({ paneRef }) {
     });
   }, []);
 
+  const sectionId = `section-${uuidv4()}`;
```
I added `useMemo` here because it doesn't seem like we want to regenerate a uuid every render

dmmulroy reviewed (approved):
Lgtm

carloskelly13 committed:
Bring back useMemo for title and uuid
